tests/dlq_test.go

```diff
@@ -0,0 +1,186 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tests
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"io"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"go.temporal.io/server/api/adminservice/v1"
+	"go.temporal.io/server/api/enums/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/tools/tdbg"
+	"go.uber.org/fx"
+)
+
+type dlqSuite struct {
+	IntegrationBase
+	*require.Assertions
+	dlq persistence.HistoryTaskQueueManager
```

Comment:
I made a task to use the ExecutableWrapper in order to write to the DLQ that way instead of directly via persistence, that way we're testing the write path and the read path in the integration test, instead of just the read path here.

Reply:
sorry I am a bit confused. The impl on L101 write the task directly to DLQ right?

Reply:
Yes. However, I think a better functional test would write to the DLQ less invasively. For example, it should make a workflow that causes task failures that cause the task to be written to the DLQ. However, since there is no way to define such a workflow, the next-best thing would be to run any workflow, and then use the ExecutableWrapper to cause executions of tasks for that workflow to produce terminal failures, which we expect task processing to send to the DLQ. We would have to make some modifications to the graph to ensure that this ExecutableWrapper is applied before the other wrappers (like the ExecutableDLQ or ExecutableDLQObserver), which is why I punted on it for now.